Add a brief guide about AST structure and metadata (#2865)
jupyter-book/mystmd: packages/myst-spec/docs/ast-primer.md

---
title: A primer on the MyST AST
short_title: AST primer
description: A getting-started guide to the structure of the MyST abstract syntax tree, for renderer and parser authors.
---

The MyST Abstract Syntax Tree (AST) is the structured representation of a MyST document after parsing.
It is plain JSON, consumed by tools that render the document (to HTML, LaTeX, Typst, etc).
This page is a brief overview of the structure and how the pieces fit together.

:::{seealso} The MyST User Guide is for users of the CLI
If you just want to author MyST documents, you probably don't need this page - head to the [MyST Guide](xref:guide).
If you want to build a renderer, write a parser, or generate AST from a plugin, read on.
:::

## The spec and mystmd

This primer covers both **the MyST Specification** (this site) and **mystmd** (the official reference implementation):

- The **spec** defines what a conforming MyST parser must produce - node types, their properties, and how they correspond to source. It evolves through the [MyST Enhancement Proposal process](https://mep.mystmd.org).

> - **mystmd** implements the spec, then goes further: it runs transforms that resolve cross-references and apply numbering, emits additional node types that aren't yet in the spec, and applies its own conventions (like named parts of a document).

Reviewer: Would it make sense to characterize `mystmd` as an _application_ vs. a parser? (Do you think of it that way?)

I wonder if that would help motivate these additional transforms that `mystmd` does. That's how I've been thinking of it—`mystmd` does this additional stuff because it is, among other things, a static site generator on top of just a MyST parser.

Reviewer 2: Yes - very much think of mystmd as a command-line application that does a lot more than parsing! We also ship `myst-parser` as one of the packages, and `myst-transforms` which implement a lot of the transformations that are inside of the application.

Author: added some more language to clarify this

If you're writing a parser, stick to the spec. mystmd's conventions (like specific named parts) aren't part of the spec, but they're still useful to know, since most MyST content today comes from mystmd. This page flags which is which as we go.

## The shape of a node

Every node in the AST is a JSON object with a `type` field that names what kind of node it is.
Beyond `type`, a node has either a `value` or `children`:

> - **Leaf nodes** have a `value` (typically a string).

Reviewer: Is this true? I guess I don't know what's actually in `mystmd` but I can think of at least one leaf node without a `value` field according to the available node type index: A link definition has a `label`, `url`, and `title` but no `value` field

Reviewer 2: A link almost always has `children` (text nodes, which are the words shown on the page that you click on).

The only `leaf` nodes that I can think of that doesn't have a `value` is an iframe and an image. So that actually might be worth updating here.

> - **Leaf nodes** have a `value` (typically a string).

Reviewer 2: ```suggestion
- **Leaf nodes** typically have a `value` (typically a string).
```

- **Parent nodes** have a `children` array (more nodes).

Nodes may also carry additional properties - `url` on a link, `depth` on a heading, `identifier` on a cross-reference, etc.
These properties are specific to each node type and are documented in the [node reference](./myst.schema.md).

> Here is the smallest meaningful AST: a single line of text.

Reviewer 3: you mean the smallest meaningful MyST, right ?

Reviewer 2: The below example is interactive, so will show the ast if you click on it... and that is explained below!

Author: tried to clarify this a bit

```{myst}
Hello, world!
```

Click the **AST** tab above to see the AST. (The tab displays YAML for readability; MyST sites serve it as JSON.)

> You'll find a `root` node (always the outermost node) wrapping a `paragraph`, which wraps a `text` leaf.

Reviewer 3: maybe already mention here that https://mystmd.org/sandbox allows to see the AST of any myst fragment ?

Author: Added mention of the sandbox earlier

## Inline formatting nests parent nodes

Inline formatting introduces parent nodes around the text they decorate.

```{myst}
A paragraph with **bold** and *emphasis*.
```

In the AST tab, the paragraph's `children` is a flat sequence of `text`, `strong`, and `emphasis` nodes. Each `strong` and `emphasis` wraps its own `text` node.
The markup characters (`**`, `*`) are gone from the source text. They've been turned into nested nodes (e.g. `*word*` becomes a `type: emphasis` node wrapping a `text` node).

(pre-and-post)=
## Two stages: PRE and POST

> In the `myst-cli`, a MyST document is processed in two phases, and the AST has a different shape after each one.

Reviewer 2: ```suggestion
In the `myst-cli`, a MyST document is processed through a series of transformations, and the AST has a different shape after each one.
```

> The short-hand for this is `PRE`- and `POST`-transforms.

Reviewer 2: ```suggestion
The short-hand for this in the online demo is `PRE`- and `POST`-transforms. You can get very granular by applying various [`myst-transforms`](https://npmjs.com/package/myst-transforms) in your own libraries or packages.
```

1. **PRE**: the output of *parsing*. The parser walks the source and expands directives and roles by calling their `run` functions. Cross-references still look like ordinary `link` nodes (e.g. `url: "#some-label"`). Directives appear as `mystDirective` wrapper nodes containing their expanded children. Numbering has not yet been applied.

> 2. **POST**: the output after [transforms](xref:guide#develop:transforms) run over the PRE tree. Link nodes pointing at labels become `crossReference` nodes with resolved targets. Most directive wrappers are unwrapped, leaving only their semantic children. Numbering and enumeration are applied. This is the form most renderers consume.

Reviewer: I'm wondering whether, given this definition of "POST," it's really a concept that is relevant to document engines other than the reference one.

What "PRE" means is very clear. It's the shape of the AST required by the spec test cases. "POST" seems to mean "this grab-bag of transforms that happen to currently be implemented by `mystmd` before rendering."

I could imagine applications that want to do some but not all of these POST transforms. Or want to do additional transforms specific to the application before rendering. [This](https://github.com/sinclairtarget/aweigh), for example, is a silly little application that adds the anchor emoji at the beginning of all link text in a MyST document. Adding the emoji is a "transform" on the AST (at least, I think of it that way), and it's done before rendering, but does that mean the AST is a "POST" AST? Or does it not count because a transform was done that `mystmd` doesn't do?

Just throwing out a wild idea: Could it make more sense to instead define the "spec-compatible" AST (what we call "PRE" now) and then certain "canonical transforms" one can do on the tree, which might also have their own test cases in the spec? Whether or not other document engines do these transforms is up to them, but if they are done, they must be done in the same way `mystmd` does them.

Author: I tried to clarify the difference between PRE and POST along with the additions that @rowanc1 made so hopefully this is a little bit clearer. I suspect we'll be able to improve this further, @sinclairtarget maybe you could open up a follow up PR to keep improving as you try things out?

Reviewer: Happy to do that!

You can see both stages inline: the `{myst}` directive used on this page exposes a **PRE** / **POST** toggle inside its AST tab, and so does the [MyST sandbox](https://mystmd.org/sandbox). Try it on this cross-reference:

```{myst}
(my-label)=
**A labeled paragraph.**

See [this paragraph](#my-label) for context.
```

In **PRE**, the second paragraph contains an ordinary `link` node with `url: "#my-label"`.
In **POST**, that same node has become a `crossReference` with `identifier`, `kind`, and `resolved: true` properties.

### Which stage will you encounter?

- The **spec's [test cases](https://unpkg.com/browse/myst-spec/dist/)** describe the **PRE** form. They are the canonical "source -> AST" pairs that any conforming parser must produce.
- The [node reference](./myst.schema.md) documents node *types* that may appear in either stage. Some properties (like `resolved` on `crossReference`, or computed numbering fields) appear only after transforms.

In short:

- A parser produces PRE. Match the test cases.
- A renderer receives POST. It's what `myst build` emits.

> - A transform takes PRE-like input and produces POST-like output.

Reviewer 3: naive/stupid question surely, but how can we invoke/chain several transforms then if input and output are of different types ?

Reviewer 2: ```suggestion
- A transform pipeline applies a series of transformations that change the tree from `PRE` to `POST`
```

Maybe this is more clear, that it starts with PRE, applies a series of transformations, and then those show something closer to the POST (depending on which transforms you applied...).

## Directives in PRE vs. POST

Directives change shape between PRE and POST (roles behave similarly):

```{myst}
:::{note}
Heads up!
:::
```

> In **PRE**, you'll see a `mystDirective` wrapper (name: `note`) containing an `admonition` child. The wrapper records the directive that produced it. The child is what its `run` function returned.

Reviewer: The reference here to a `run` function seems unnecessarily specific to the existing implementation

Reviewer 2: ```suggestion
In **PRE**, you'll see a `mystDirective` wrapper (name: `note`) containing an `admonition` child. The wrapper records the directive that produced it. The child is what its `run` function returned from a `mystmd` plugin.
```

In **POST**, the `mystDirective` wrapper is typically stripped, and transforms may add extra structure. For example, [`admonitionHeadersTransform`](https://github.com/jupyter-book/mystmd/blob/0d626b198093fa740125d02267a111d84547fd36/packages/myst-transforms/src/admonitions.ts#L36) inserts an `admonitionTitle` child.

When directive parsing fails, the POST AST signals it in one of two ways:

- **Unknown directive name** (no plugin registered): the node stays as a `mystDirective` leaf, with the original source text in its `value`.
- **Invalid arguments or body**: the node's type becomes `mystDirectiveError`, and its children are dropped. Renderers can use this to display error messages in the output.

## Plugins often emit a custom node that is transformed

A common plugin pattern uses a directive (or role) to emit a *placeholder* node, then a separate transform to expand that placeholder into real content.

> This is useful for anything you can't do inside `run()`: async work, anything that needs the whole document for context, or anything that depends on cross-references being resolved first.

Reviewer 3: I did not get that sentence; what does `run()` refer to exactly ?
in any case I'd suggest to move this below the example

Reviewer 2: The `run` function is referring to mystmd plugins here:
https://mystmd.org/guide/javascript-plugins#defining-a-new-directive

We could update that to make it more clear...

For a working in-repo example of this pattern, see [`docs/templates.mjs`](https://github.com/jupyter-book/mystmd/blob/main/docs/templates.mjs) (the `myst:template` directive emits a `myst-template-ref` placeholder, then `mystTemplateTransform` fetches template metadata and expands them).

## Frontmatter and document structure

A document's frontmatter (the YAML at the top of a `.md` file) doesn't appear as a node in the tree.
It is attached separately, alongside the `mdast` (the AST) in the JSON output of `myst build`.
A built page on disk looks roughly like:

```{code} json
:caption: A page from `myst build --site`
{
  "kind": "Article",
  "frontmatter": { "title": "...", "authors": [...] },
  "mdast": { "type": "root", "children": [ ... ] },
  "references": { ... }
}
```

The `mdast` field is the tree you've been inspecting in the AST tabs above.

> The `references` field holds cross-document data like resolved cross-references and citations, which a renderer needs when working across pages.

Reviewer: I'm a little unclear from this section what the relationship is between the AST, the spec, and this JSON structure for a built `mystmd` page.

Frontmatter is not a part of the MyST spec, right? And this JSON structure seems specific to the concerns of `mystmd` the application / static site builder. Presumably a MyST parser wouldn't have to worry about it?

Author: I've tried to clarify the relationship between frontmatter/parts and the AST a bit more - that stuff is now documented in the mystmd guide rather than the spec!

## Chunking a page into blocks

The `+++` syntax divides a page into `block` nodes. Each `block` wraps the content that follows it, and optional JSON after `+++` becomes block metadata:

```{myst}
+++ {"label": "intro", "foo": "bar"}

A first chunk.

+++

A second chunk.
```

In the AST (POST), each chunk ends up inside a `block` node. Recognized keys (`kind`, `class`, `label`, `identifier`) are moved onto the `block` node, anything else is put under `block.data`. See [](./features/blocks.md) for the formal block specification.

## "Parts": named regions of a page

> 

Reviewer 3: disclaimer: I have note been able to follow this section on parts (have never been exposed to that..)

Reviewer 2: @choldgraf I think we could safely delete the parts explainer from here, I think it is beyond the AST-primer and getting into mystmd specifics potentially...

Author: OK cool - let me see if there's a natural space for this in the mystmd guide docs and we can cross-reference it as an example of something that's more myst-cli rather than myst-spec

Author: I've moved the discussion of parts to the mystmd guide, so that it's just mentioned briefly in the spec and then cross-referenced there
